src/machine_learning/optimization/k_nearest_neighbors.rs

```diff
@@ -0,0 +1,294 @@
+#![allow(dead_code)]
+#![allow(unused_variables)]
+
+use std::{
+    cmp,
+    collections::{BTreeMap, HashMap},
+};
+
+use crate::math::{abs, square_root};
+#[derive(Debug, Clone)]
+pub struct Point {
+    x: f64,        // x-axis
+    y: f64,        // y-axis
+    label: String, // A label denotes the classification of the data. For instance, 'Football' might be labeled as 'Sport'
+}
+
+impl Point {
+    pub fn build(x: f64, y: f64, label: String) -> Self {
+        Self { x, y, label }
+    }
+}
+
+pub enum DistanceMeasurementFormula {
+    EucledianDistance, // The Euclidean Distance Formula will be employed to calculate the distance between two points
+}
+
+impl DistanceMeasurementFormula {
+    fn distance(&self, source_point: &Point, destination_point: &Point) -> f64 {
+        match self {
+            Self::EucledianDistance => Self::eucledian_distance(source_point, destination_point),
+        }
+    }
+
+    fn eucledian_distance(source_point: &Point, destination_point: &Point) -> f64 {
+        let distance = square_root(
+            (destination_point.x - source_point.x) * (destination_point.x - source_point.x)
+                + (destination_point.y - source_point.y) * (destination_point.y - source_point.y),
+        );
+        abs(distance)
+    }
+}
+
+pub fn classify_using_knn(
+    neighbors: Vec<&Point>, // The training data which essentially consists of a set of points on the X-Y axis represented as vector of Points,
+    input_point: Point,     // The input point requiring classification
+    k: usize, // The value of 'K'. For example, if K equals 4, classification is determined by the majority vote among the 4 nearest neighbors
+    distance_computation: DistanceMeasurementFormula, // An enum employed to specify the technique/formula for calculating the distance between two points
```

Comment:
Why not do make it a _function pointer_, so the caller could define their own _distance_ without modifying your code?

Reply:
Yeah, I was also thinking about using traits to achieve the same. But i assume, traits do add a vtable's lookup overhead which can be avoided with function pointers. I updated the code with the same